### ISSUE TITLE
usbmux breaks Touch Bar, ALS, and camera on MacBookPro's

### ISSUE THREAD

**Issue by roadrunner2** [libimobiledevice/usbmuxd#138]
When running on a MacBookPro, commits ec5ff91cfabd30637f8af8f5c79baf4d7818ce57 and 8a69f1a78a58476f77b66916091d2405d0cd815f cause usbmux to misconfigure (configuration 0 is not a valid config) the iBridge device (T1 chip), breaking Touch Bar, light sensor, and camera functionality. The problem was originally reported at roadrunner2/macbook12-spi-driver#42.

I don't really understand the usbmux in any detail, so not sure how to fix this exactly, but it seems like you need to distinguish between internal USB devices and external ones plugged into the system, and avoid touching the internal ones.

**Comment by mexmer** [comment]
usbmuxd is not supposed to be run on OSX, it has own mux server

**Comment by nikias** [comment]
@mexmer I think he is running it on Linux on the MacBook Pro.

**Comment by mexmer** [comment]
oh i see.

now question is, how to limit usbmux only to ios devices, without need to explicitly list them.

**Comment by nikias** [comment]
@mexmer I am working on it with @rickmark.

**Comment by nikias** [comment]
On Linux, configuration 0 basically means the device is "unconfigured". Not sure why this breaks though; the kernel driver should know about this. Anyway, I think maybe the 8600 device might just need a different rule. Can you guys tell me what the "normal" config of the T2 is? Does it work after modifying the udev rule to set it to 1 on the 8600 (T2)? Try this patch:

```diff
diff --git a/udev/39-usbmuxd.rules.in b/udev/39-usbmuxd.rules.in
index ab20517..7f9bf10 100644
--- a/udev/39-usbmuxd.rules.in
+++ b/udev/39-usbmuxd.rules.in
@@ -4,7 +4,8 @@
 SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/190[1-4]/*|5ac/8600/*", TAG+="systemd"

 # Initialize iOS devices into "deactivated" USB configuration state and activate usbmuxd
-SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/190[1-4]/*|5ac/8600/*", ACTION=="add", ENV{USBMUX_SUPPORTED}="1", ATTR{bConfigurationValue}="0", OWNER="usbmux", @udev_activation_rule@
+SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/190[1-4]/*", ACTION=="add", ENV{USBMUX_SUPPORTED}="1", ATTR{bConfigurationValue}="0", OWNER="usbmux", @udev_activation_rule@
+SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/8600/*", ACTION=="add", ENV{USBMUX_SUPPORTED}="1", ATTR{bConfigurationValue}="1", OWNER="usbmux", @udev_activation_rule@

 # Make sure properties don't get lost when bind action is called
 SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/190[1-4]/*|5ac/8600/*", ACTION=="bind", ENV{USBMUX_SUPPORTED}="1", OWNER="usbmux", @udev_activation_rule@
```

/cc @davidjo

**Comment by davidjo** [comment]
For quick fix just copy the 39-udevadm.rules file from Ubuntu 20.04 (which is missing all |5ac/8600 additions) and touchbar is restored.

**Comment by davidjo** [comment]
I just tried this - but get a syslog error
systemd-udevd[353]: /usr/lib/udev/rules.d/39-usbmuxd.rules:7 Invalid key/value pair, ignoring.
systemd-udevd[353]: /usr/lib/udev/rules.d/39-usbmuxd.rules:8 Invalid key/value pair, ignoring.

lines 7 8 are the green lines in your patch

of course the message doesnt say which key is the problem

(so touchbar works because udev doesnt understand the file so ignores it)

**Comment by night199uk** [comment]
> On Linux, configuration 0 basically means the device is "unconfigured". Not sure why this breaks though; the kernel driver should know about this. Anyway, I think maybe the 8600 device might just need a different rule. Can you guys tell me what the "normal" config of the T2 is? Does it work after modifying the udev rule to set it to 1 on the 8600 (T2)? Try this patch:
> 
> ```diff
> diff --git a/udev/39-usbmuxd.rules.in b/udev/39-usbmuxd.rules.in
> index ab20517..7f9bf10 100644
> --- a/udev/39-usbmuxd.rules.in
> +++ b/udev/39-usbmuxd.rules.in
> @@ -4,7 +4,8 @@
>  SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/190[1-4]/*|5ac/8600/*", TAG+="systemd"
> 
>  # Initialize iOS devices into "deactivated" USB configuration state and activate usbmuxd
> -SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/190[1-4]/*|5ac/8600/*", ACTION=="add", ENV{USBMUX_SUPPORTED}="1", ATTR{bConfigurationValue}="0", OWNER="usbmux", @udev_activation_rule@
> +SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/190[1-4]/*", ACTION=="add", ENV{USBMUX_SUPPORTED}="1", ATTR{bConfigurationValue}="0", OWNER="usbmux", @udev_activation_rule@
> +SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/8600/*", ACTION=="add", ENV{USBMUX_SUPPORTED}="1", ATTR{bConfigurationValue}="1", OWNER="usbmux", @udev_activation_rule@
> 
>  # Make sure properties don't get lost when bind action is called
>  SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/190[1-4]/*|5ac/8600/*", ACTION=="bind", ENV{USBMUX_SUPPORTED}="1", OWNER="usbmux", @udev_activation_rule@
> ```
> 
> /cc @davidjo
> 
> Don't forget to rebuild and reinstall so the udev rule gets updated.

This patch actually seems to work great. I see it's been a while since this bug was updated. Any chance this patch could get integrated, as with Touch Bar support now in the mainline kernels this would fix the OOTB Touch Bar experience on most T1/T2 Macs.

**Comment by davidjo** [comment]
could you post the actual 39-usbmuxd.rules here - hadnt realized this is a pre-cursor file which is filled in and setting up the makefile seems complicated as it just replaces the udev_activation_rule parameter with other parameters which is not clear where they are defined.
(This explains my symbol issues).

**Comment by KiLLeRRaT** [comment]
I've also run the patch @nikias posted, and worked for me too.

It'd be create to get this merged in.

Cheers,

**Comment by davidjo** [comment]
could you post the value of the string @udev_activation_rule@ in your  actual udev file please

**Comment by KiLLeRRaT** [comment]
Sure thing, here you go:

`cat /usr/lib/udev/rules.d/39-usbmuxd.rules`:
```
# usbmuxd (Apple Mobile Device Muxer listening on /var/run/usbmuxd)

# systemd should receive all events relating to device
SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/8600/*", TAG+="systemd"

# Initialize iOS devices into "deactivated" USB configuration state and activate usbmuxd
SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/190[1-4]/*", ACTION=="add", ENV{USBMUX_SUPPORTED}="1", ATTR{bConfigurationValue}="0", OWNER="usbmux", @udev_activation_rule@
SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/8600/*", ACTION=="add", ENV{USBMUX_SUPPORTED}="1", ATTR{bConfigurationValue}="1", OWNER="usbmux", @udev_activation_rule@

# Make sure properties don't get lost when bind action is called
SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/8600/*", ACTION=="bind", ENV{USBMUX_SUPPORTED}="1", OWNER="usbmux", ENV{SYSTEMD_WANTS}="usbmuxd.service"

# Exit usbmuxd when the last device is removed
SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ENV{PRODUCT}=="5ac/12[9a][0-9a-f]/*|5ac/8600/*", ACTION=="remove", RUN+="/usr/bin/usbmuxd -x"
```

**Comment by nikias** [comment]
Fixed with 61b99ab5c25609c11369733a0df97c03a0581a56.
